Added close handler on initial response for reactive SseEventSinkImpl

a29340 opened the pull request:
This pr adds the necessary handling of the close event on SseEventSinkIpl. This addresses the scenario when it is the client that is closing the connection, and the sink was not registering itself as closed and most importantly the broadcaster was never notified of the sink close state.

a29340 committed:
addressing resteasy reactive SseBroadcaster issues

gsmet commented:
@a29340 no need to rebase for now. @FroMage is on PTO for now. I will ping him when he's back so he can have a look to this one and the other SSE-related one.

a29340 commented:
> @a29340 no need to rebase for now. @FroMage is on PTO for now. I will ping him when he's back so he can have a look to this one and the other SSE-related one.

Oh ok, thank you!

rjvq85 commented on independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/jaxrs/SseEventSinkImpl.java:
```diff
@@ -75,10 +75,13 @@ public void accept(Throwable throwable) {
-            //                // FIXME: notify of client closing
-            //                System.err.println("Server connection closed");
-            //            });
+            response.addCloseHandler(() -> {
```
Hi!

I checked out your branch and was testing the solution but I cannot understand why the close handler is only added when `!response.headWritten()`. Is this on purpose? For some reason, on my application this condition is always false and because of that the Close event firing never happens.

a29340 replied:
Hi @rjvq85, thanks for taking a look! 
No, this was not placed there with a specific purpose, I just added the closeHandler where the previous comment was, but it makes sense to move it out of the if statement, to always add it to the response. 

For me the condition `!response.headWritten()` was always true when testing, so it always execute the close handler, and I cannot reproduce a case where it gets evaluated to false... I used the reproducer provided by #23997, with the `sse-server-reactive` module and it worked fine, the issues on the reactive server seemed to be resolved.

I tested again after moving the changes out of the if statement, and I don't see any unexpected behaviour, so i guess it's fine like that?

FroMage commented:
Also, could you add a test for this fix?

a29340 commented on independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/jaxrs/SseBroadcasterImpl.java:
```diff
@@ -72,7 +72,9 @@ public synchronized void close() {
     synchronized void fireClose(SseEventSinkImpl sseEventSink) {
         for (Consumer<SseEventSink> listener : onCloseListeners) {
-            listener.accept(sseEventSink);
+            if (sinks.remove(sseEventSink)) {
```
I don't get what do you mean, maybe I'm missing something. If we don't remove the sink (which is closing) from the broadcaster, that sink will continue to receive events even if it is closed, right?

FroMage replied:
You have `n` listeners. On the first one, you remove the sink, then it's removed, and for all remaining listeners the sink is already removed so you don't notify them. That's a logic bug, no?

a29340 replied:
Oh, now I see it, sorry! You're right, the removal should happen after all the listeners have been notified. I'll convert this pr to draft and work some more on it, since I'm already adding some tests.

a29340 commented on independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/jaxrs/SseEventSinkImpl.java:
```diff
@@ -75,11 +75,14 @@ public void accept(Throwable throwable) {
-            //            });
         }
+        response.addCloseHandler(() -> {
+            if (broadcaster != null) {
```
That works if the server is the one that closes the connection, the `isClosed()` returns `false` so the execution continues and `broadcaster.fireClose(this)` at line 59 is actually called.  When the client is the one that drops, `isClosed()` method already returns `true`  when the closeHandler is executed and the broadcaster is never notified. 
At least this was the case using the reproducer [here](https://github.com/jimbogithub/sse) (using reactive-server)

a29340 commented:
Hi @FroMage,

I added mockito as a dependency to do some tests, is it ok?

FroMage commented:
Thanks. I'm generally in favour of real HTTP tests rather than Mockito. Do you think you could write a `RestAssured` test that exhibits the issue you're fixing?

gsmet commented:
@a29340 any chance you could rebase and finalize this?

I don't think we have fixed this issue already?

a29340 commented:
@gsmet Yes, I will try to finalize this in the next few days. Sorry, I didn´t have a chance to work on it recently

FroMage reviewed (changes requested):
LGTM, could you rebase on the latest `main`?

a29340 commented:
I'm moving this to draft again, some tests are still unstable (often they fail) when run on the CI pipeline on the forked branch.

a29340 commented:
Hi @FroMage, sorry for the long wait. I finally did fix up the tests covering this PR, and rebased on the latest `main` and squashed in a single commit

a29340 commented:
Hi @gsmet @FroMage, is this PR still desired? Should I rebase again waiting for a review?

FroMage commented:
> Hi @gsmet @FroMage, is this PR still desired? Should I rebase again waiting for a review?

Yes please. Sorry about the delays :(

a29340 committed:
updated mockito and bytebuddy to support java 20

FroMage commented on independent-projects/resteasy-reactive/pom.xml:
```diff
@@ -49,7 +49,7 @@
         <jakarta.enterprise.cdi-api.version>4.0.1</jakarta.enterprise.cdi-api.version>
         <jandex.version>3.1.5</jandex.version>
-        <bytebuddy.version>1.12.12</bytebuddy.version>
+        <bytebuddy.version>1.14.6</bytebuddy.version>
```
This should be aligned with the one used by the Quarkus BOM, required by Hibernate ORM.

a29340 replied:
updated right now to 1.14.7, that is defined in quarkus BOM

FroMage reviewed (changes requested):
Sorry for the late review. This is OK, except the byte-buddy dependency that should be aligned with the Quarkus BOM.

FroMage commented:
The first test failure seems related to this fix, no?

a29340 commented:
> The first test failure seems related to this fix, no?

Yes, looks like the test didn't detect the expected call on close method on the jdk 20 build. I'm trying to replicate on a local environment, but the tests pass every time. On the forked repo the [jdk 11 build failed instead](https://github.com/a29340/quarkus/actions/runs/6455749875/job/17524617495), with the same reason as this build, but passed on the 2nd attempt. Looks like it's not related to the specific java version. 
I'll try to make this test more stable and run again a couple of times on the fork before requesting a build on here too

a29340 committed:
updated mockito version

removed bom for mockito test dependency

sorted imports

removed colors and added logger instead of system out println

formatted

added more logs

added more logs

added more logs, updated bytebuddy

formatted

adding details to test log

FroMage commented on independent-projects/resteasy-reactive/pom.xml:
```diff
@@ -97,6 +98,15 @@
+                <type>pom</type>
+                <scope>import</scope>
+            </dependency>
+
```
Just want to double check with @geoand and @gsmet that they're OK with this dependency being added.

geoand replied:
Do we really need the BOM? Isn't adding the mockito test dependency enough?

FroMage replied:
🤷

a29340 replied:
actually no, let me remove it. I'll try tu run the ci on the fork and then rebase

FroMage reviewed (changes requested):
Temporarily asking for changes, just to make sure we don't merge this before @gsmet and @geoand answer the dependency question. But my review remains OK otherwise.

a29340 commented:
> > The first test failure seems related to this fix, no?
> 
> Yes, looks like the test didn't detect the expected call on close method on the jdk 20 build. I'm trying to replicate on a local environment, but the tests pass every time. On the forked repo the [jdk 11 build failed instead](https://github.com/a29340/quarkus/actions/runs/6455749875/job/17524617495), with the same reason as this build, but passed on the 2nd attempt. Looks like it's not related to the specific java version. I'll try to make this test more stable and run again a couple of times on the fork before requesting a build on here too

Hi @FroMage, I tried to replicate the same test failure on the fork ci for a while now, but couldn't find what was the issue. As of now the tests pass every time I run the ci on the fork. Can you please take a look at the tests in `SseServerTestCase` to double check?

FroMage reviewed (approved):
So, all tests pass, you figured it out?

a29340 commented:
> So, all tests pass, you figured it out?

@FroMage No, the only thing I did was rebase and add logs, but for the last month or so I couldn't replicate the test failure. As of now the tests pass reliably, but I really don't know why they didn't before

FroMage commented:
Well, we'll see if this occurs again.
